server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/CustomIdentityDAO.scala

```diff
@@ -120,15 +131,42 @@ class IdentityRepository @Inject()(customIdentityDao: CustomIdentityDAO, identit
       SMono.error(ForbiddenSendFromException(creationRequest.email))
     }
 
-  def list(user: Username): Publisher[Identity] = SFlux.merge(Seq(
-    customIdentityDao.list(user),
-    SMono.fromCallable(() => identityFactory.listIdentities(user))
-      .subscribeOn(Schedulers.elastic())
-      .flatMapMany(SFlux.fromIterable)))
+  def list(user: Username): Publisher[Identity] = {
+    val customIdentities: SFlux[(String, Identity)] = SFlux.fromPublisher(customIdentityDao.list(user))
+      .map(e=> (IdentityOrigin.CUSTOM, e))
 
-  def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit] = customIdentityDao.update(user, identityId, identityUpdate)
+    val serverSetIdentities: SFlux[(String, Identity)] = SMono.fromCallable(() => identityFactory.listIdentities(user))
+      .subscribeOn(Schedulers.elastic())
+      .flatMapMany(SFlux.fromIterable)
+      .map(e=> (IdentityOrigin.SERVER_SET, e))
+
+    SFlux.merge(Seq(customIdentities, serverSetIdentities))
+      .groupBy(_._2.id)
+      .flatMap(_.reduce((id1, id2) => id1._1 match {
+        case IdentityOrigin.CUSTOM => id1
+        case _ => id2
+      }))
```

Comment:
IMO we should have an `identity.merge(otherIdentity)` method.

Reply:
We have no field in Identity for priority checking. IMO, we shouldn't wrap Identity or add a new field for only this logic?

Reply:
This code block is hard to read and would benefit from being extracted. Or simplified.